hashcracker/cracker/lib/cracker/dispatcher.ex

```diff
@@ -0,0 +1,118 @@
+defmodule Cracker.Dispatcher do
+  use GenServer
+
+  @brute_force_upper_bound 64
+
+  ##################################################
+  # External API
+  ##################################################
+
+  def start_link do
+    GenServer.start_link(__MODULE__, [], name: __MODULE__)
+  end
+
+  def start(state=%{attack: mode}) do
+    GenServer.cast(__MODULE__, {mode, state})
+  end
+
+  def not_found(pid) do
+    GenServer.cast __MODULE__, {:not_found, pid}
+  end
+
+  def found_pass(pass) do
+    GenServer.cast __MODULE__, {:found_pass, pass}
+  end
+
+  ##################################################
+  # GenServer implementation
+  ##################################################
+
+  def handle_cast({:not_found, worker}, state) do
+    state = worker_done(worker, state, Enum.count(state.workers))
+    {:noreply, state}
+  end
+
+  def handle_cast({:found_pass, pass}, state) do
+    Supervisor.stop(Cracker.DispatcherSupervisor)
+    send state.client_pid, {:pass_found, pass}
+    {:noreply, state}
+  end
+
+  def handle_cast({_mode, state}, _) do
+    workers = start_workers(state.num_workers)
+    state = Map.put(state, :workers, workers)
+    dispatch(state)
+    {:noreply, state}
+  end
+
+  ##################################################
+  # Implementation
+  ##################################################
+
+  defp start_workers(num_workers) do
+    {:ok, _} = Cracker.DispatcherSupervisor.start_link(num_workers)
+    Supervisor.which_children(Cracker.DispatcherSupervisor)
+    |> Enum.map(fn {_, pid, _, _} -> pid end)
+  end
+
+  defp split_mask_head_tail(mask) do
+    case String.split(mask, "?", trim: true, parts: 2) do
+      [ h | [ mask ] ] ->
+        {h, mask}
+      [ h | [] ] ->
+        {h, ""}
+    end
+  end
+
```

Comment:
I wonder if the strategy could be factored away from the housekeeping here. All three functions end with:

~~~ elixir
+        state
 +        |> Map.drop([:workers])
 +        |> Map.merge(%{worker: worker, start: start, chunk_size: chunk_size})
 +        |> Cracker.Worker.start_work
~~~

This makes me think that there could be one dispatch function that chooses one of three other functions to do the strategy stuff, and then handles the worker setup.